### ISSUE TITLE
compatibility with Plan 9 rc: `if not`

### ISSUE THREAD

**Issue by GReagle** [rakitzis/rc#19]
Greetings.  It would be great if scripts could be written that would run in both Duff's rc and this (Rakitzis's) rc.

To achieve this, the author of such a script needs to avoid extra features in this rc (e.g. `return`, `break`), and extra features in Duff's rc (e.g. `flag`).  However, the `else` versus `if not` difference seems to be a major incompatibility.

Can this rc be modified to accept `if not`?

**Comment by rakitzis** [comment]
I took the cue from the paper:


```
Rc’s admittedly feeble solution is to declare that the else clause is a separate statement, with the semantic proviso that it must immediately follow an if, and to call it if not rather than else, as a reminder that something odd is going on.
```

as a hint to improve on "if not". I can't imagine changing it at this point, 25 years in.

**Comment by GReagle** [comment]
I want to make sure that you understand that I am proposing **adding** the ability to accept `if not`.  I am **not** proposing removing the ability to accept `else`.

Wow!  25 years is a long time.  Congratulations!

**Comment by rakitzis** [comment]
Right: the "if not" handling is also an accident of history. Plan 9 was well-protected by corporate AT&T. There was no sense that it would ever see the outside, so I sat down to reimplement the shell for Unix because there was simply no alternative at the time. Practically the entire suite of shells was: sh, csh, ksh   and bash (still quite new).

**Comment by TobyGoodwin** [comment]
I'm broadly agreeable to making minor tweaks to `rc` that will make life easier for people in the real world. In this case, we'd be appropriating what's currently invalid syntax to make it much easier to write a script that will run on both the Duff and Rakitzis versions of  `rc`

I don't have ready access to Plan 9 a Duff `rc`. Online man pages state:

> The immediately preceding command must have been `if(list) command`

but a quick perusal of the sources on http://9p.io/ suggests to me that this is not enforced. (It's entirely possible I was looking in the wrong place though!) Can somebody check this for me?

**Comment by TobyGoodwin** [comment]
There's a branch `ifnot` that implements this, although it comes with a very large **warning** that it has been extremely lightly tested.
Mostly, this treats `if not` as syntactic sugar. When parsing, we may generate `nIfnot` nodes in the tree, which are subsequently removed if they are in a valid location, i.e. after an `nIf` inside an `nBody`.
As far as I can tell (bearing in mind the warning above), the only other case is at the top level of an interactive shell. So there's a flag in `walk.c` to handle this correctly. This means that you can copy an `if` / `if not` from a script, paste it into a terminal, and it will do the Right Thing.
Comments welcome, especially from anybody who gets a chance to test it.

**Comment by muennich** [comment]
Is this supposed to work yet?
```
; if (true) echo ja; if not echo nein
here i am!
ja
rc: `if not' must follow `if'
```

**Comment by TobyGoodwin** [comment]
dammit, i think i've messed up a merge somewhere :(

will fix asap

**Comment by TobyGoodwin** [comment]
After spending some time with `git fsck` and friends, I have failed to find the code I know I wrote :( Have now reimplemented it, which is probably just as well as I noticed I'd missed an important case.

This time it's definitely committed: 7a643120366f1ea24eac5ccb759d4df616e7789b

As before, this is “lightly” tested - please do try to break it!

**Comment by TobyGoodwin** [comment]
Thanks @borkovic for your persistence with this, and especially for the test which I've incorporated.

I've pushed a new implementation that I think is a little bit neater.

I'm not surprised by your discovery that it matters whether `if not` is on the same line. Have you figured out why yet? (Hint: `walk.c`! :smile:)

As always, please test and try to break.

@GReagle as the original requester, do you have any comments on what we have so far? Do you have any scripts written for the Duff rc that you could test with the latest version?

**Comment by TobyGoodwin** [comment]
Umm, sorry for the duff hint! I meant `tree.c`

**Comment by GReagle** [comment]
I don't have the time right now to compile and test this enhancement to Rakitzis rc.  Maybe in the future.  However, I am happy to provide a script that runs on Duff rc that uses `if not`.  It is a countdown timer script that I use occasionally.  Others feel free to use it and test it.

```
#!/usr/bin/env rc
# this script assumes the GNU date command

# all numeric time variables are in seconds
start=`{date +%s}  #`
switch ($#*) {
case 0  # 5 minutes by default
	delay=300
case 1  # interpret first and only arg as minutes
	if (printf '%s' $1 | grep -Eq '^[0-9]+$') delay=`{expr $1 '*' 60}  #`
	if not {printf '%s must be a non-negative integer\n' $1; exit 1}
case 2  # interpret second arg as seconds ignoring first arg (-s)
	if (printf '%s' $2 | grep -Eq '^[0-9]+$') delay=$2
	if not {printf '%s must be a non-negative integer\n' $2; exit 1}
case *
	printf 'countdown [ <minutes> | -s <seconds> ]\n'; exit 1
}
stop=`{expr $start + $delay}  #`
printf 'Starting at ' ; date -d@$start
printf 'Stopping at ' ; date -d@$stop
current=`{date +%s}  #`
while (test $current -lt $stop) {
	rem=`{expr $stop - $current}  #`
	remp=`{date -d0+$rem' sec' +%H:%M:%S} #`
	printf \r$remp
	set-title $remp
	sleep 1
	current=`{date +%s}  #`
}
printf '\a\rDONE!   \a\n'
date
xeyes
```

**Comment by TobyGoodwin** [comment]
Thanks! I'm happy to confirm that your script works correctly in the current dev version (and of course, bombs out with syntax errors in the the 1.7.4 release).

I've also fixed another problem that @borkovic pointed out where we allowed

    if (...) { ... } else ...; if not ...

Closing this issue now, as I think we've finally knocked `if not` into a decent shape. Of course, you are all urged to re-open it (or open a new one) if you find any further problems, or want to discuss it further.
